tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_thunk.cc

```diff
@@ -109,30 +150,71 @@ Status CudnnBatchNormForwardInferenceThunk::ExecuteOnStream(
       MakeDescriptors(hlo_instruction()->shape(), feature_index_);
 
   se::DeviceMemory<float> null_device_ptr(nullptr);
-  se::DeviceMemory<float> output(buffer_allocations.GetDeviceAddress(output_));
   auto op_profiler =
       params.profiler->MakeScopedInstructionProfiler(hlo_instruction());
-  stream.ThenBatchNormalizationForward(
-      se::DeviceMemory<float>(buffer_allocations.GetDeviceAddress(operand_)),
-      se::DeviceMemory<float>(buffer_allocations.GetDeviceAddress(scale_)),
-      se::DeviceMemory<float>(buffer_allocations.GetDeviceAddress(offset_)),
-      se::DeviceMemory<float>(buffer_allocations.GetDeviceAddress(mean_)),
-      se::DeviceMemory<float>(buffer_allocations.GetDeviceAddress(variance_)),
-      /*side_input=*/null_device_ptr,
-      operand_desc,                         //
-      scale_offset_desc,                    //
-      epsilon_,                             //
-      se::dnn::ActivationMode::kNone,       //
-      &output,                              //
-      /*batch_mean=*/nullptr,               //
-      /*batch_var=*/nullptr,                //
-      /*saved_mean=*/nullptr,               //
-      /*saved_inv_var=*/nullptr,            //
-      /*is_training=*/false,                //
-      /*var_to_inv_var=*/nullptr,           //
-      /*inv_var_to_var=*/nullptr,           //
-      /*reserve_space_allocator=*/nullptr,  //
-      /*workspace_allocator=*/nullptr);
+  se::DeviceMemoryBase output_base =
+      buffer_allocations.GetDeviceAddress(output_);
+  PrimitiveType output_primitive_type =
+      hlo_instruction()->shape().element_type();
+
+  switch (output_primitive_type) {
+    case F16: {
+      auto output = se::DeviceMemory<Eigen::half>(output_base);
+      stream.ThenBatchNormalizationForward(
```

Comment:
Extract out into a template function?

Reply:
Done in https://github.com/tensorflow/tensorflow/pull/32887